lua/ccc/picker/custom_entries.lua

```diff
@@ -0,0 +1,63 @@
+local parse = require("ccc.utils.parse")
+local pattern = require("ccc.utils.pattern")
+
+---@class CustomEntries: ColorPicker
+---@field rgb { [string]: integer[] }
+---@field min_length integer
+---@field color_table { [string]: string }
+local CustomEntries = {}
+
+---@param color_table { [string]: string }
+---@return CustomEntries
+CustomEntries.new = function(color_table)
+  return setmetatable({ color_table = color_table, rgb = {}, min_length = 0 }, { __index = CustomEntries })
+end
+
+function CustomEntries:init()
+  if self.pattern then
+    return
+  end
+  ---@type { plain: string, vim: string }[]
+  local patterns = {}
+  for name, rgb in pairs(self.color_table) do
+    local r, g, b = rgb:match("(%x%x)(%x%x)(%x%x)")
+    if r then
+      local escaped = name:gsub([[\]], "%0%0")
+      table.insert(patterns, { plain = name, vim = escaped })
+      self.rgb[name] = { parse.hex(r), parse.hex(g), parse.hex(b) }
+    end
+  end
+  -- Sort names to match the longest word at first.
+  table.sort(patterns, function(a, b)
+    return #a.plain > #b.plain
+  end)
+  self.min_length = #patterns[#patterns].plain
+  local vim_patterns = vim.tbl_map(function(v)
+    return v.vim
+  end, patterns)
+  self.pattern = [[\V\<\%(]] .. table.concat(vim_patterns, [[\|]]) .. [[\)\>]]
```

Comment:
I don't really see the advantage of storing in `patterns` once, sorting, tbl_map. It's not even easy to read. Why not just do it in one loop?

```lua
function CustomEntries:init()
  if self.pattern then
    return
  end
  ---@type string[]
  local vim_patterns = {}
  ---@type integer
  local min_length
  for name, rgb in pairs(self.color_table) do
    local r, g, b = rgb:match("(%x%x)(%x%x)(%x%x)")
    if r then
      self.rgb[name] = { parse.hex(r), parse.hex(g), parse.hex(b) }
      local escaped = name:gsub([[\]], "%0%0")
      table.insert(vim_patterns, escaped)
      if min_length == nil or min_length > #name then
        min_length = #name
      end
    else
      vim.notify("Invalid color expression:", rgb)
    end
  end
  self.pattern = [[\V\<\%(]] .. table.concat(vim_patterns, [[\|]]) .. [[\)\>]]
  self.min_length = min_length
end
```

Reply:
Good point. Yes, it is enough to loop once. I will fix this.

Reply:
Ah, I have remembered the reason I implemented this current logic.

Let's say, we want to colorize `hoge` and `hoge-fuga` into different colors.

```lua
ccc.setup {
  pickers = {
    ccc.picker.custom_entries {
      hoge = "#ff0000",
      ["hoge-fuga"] = "#00ff00",
    }
  }
}
```

The logic you wrote makes regexp: `\V\<\%(hoge|hoge-fuga\)\>`. Then, when you write `hoge & hoge-fuga` in one line, it highlights only `hoge`'s.

```lua
hoge & hoge-fuga
-- ↑ colorized like this
-- {hoge} & {hoge}-fuga
-- ↑ {} means it is colored as #ff0000.
```

This is because Vim returns the captured characters as soon as it finds the ones -- `hoge`, not `hoge-fuga`. So we should make regexp with candidates sorted reversely.

```lua
-- the original table
{
  hoge = "#ff0000",
  ["hoge-fuga"] = "#00ff00",
  ["hoge-fuga-egoh"] = "#0000ff",
}

-- regexp we should make
[[\V\<\%(hoge-fuga-egoh|hoge-fuga|hoge)\>]]
```

Yes, this is an arbitrary example. `hoge-fuga` is not a “word”, but it is possible that users will set here more than “words”. I think this logic is needed here.

Reply:
-- If more than one branch matches, the first one is used.
-- Since key is not guaranteed to be iskeyword, order from longest to shortest.